org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/project/registry/ProjectRegistryManager.java

```diff
@@ -917,6 +918,9 @@ static class Context {
 
     request.setLocalRepository(getMaven().getLocalRepository());
     request.setWorkspaceReader(getWorkspaceReader(state, pom, resolverConfiguration));
+    if(pom != null && pom.getLocation() != null) {
+      request.setMultiModuleProjectDirectory(MavenImpl.computeMultiModuleProjectDirectory(pom.getLocation().toFile()));
```

Comment:
If we know the pom why have all the basedir hassle above? It seems this is only required here, because if the request was not aquired by a project then the `MultiModuleProjectDirectory` won'T have a real meaning

Reply:
Actually you can use the IProject location, also pomless *has* poms if we would use more maven like execution and if we would support extensions, what I try to in cooperate e.g. with #688 ... so from maven POV there is always a pom but not necessarily named `pom.xml` so one could safely assume that such a `MultiModuleProjectDirectory` would require some kind of "entry-point".

Reply:
> We don't need an IProject to get an IMavenProjectFacade, and it's important that things remain as such.

Okay then maybe `IMavenProjectFacade.getProject()` gives a wrong impression here, anyway I assume that each `IMavenProjectFacade` has a directory (aka basedir?) assigned, so this should be the entry point (a pom.xml is not needed). So then `IMavenExecutionContext.getExecutionRequest()` should contain a variant where one can pass a `IMavenProjectFacade` then which is then used to construct the appropiate Execution context without the need to store the basedir in the IMaven.

Reply:
> IMavenProjectFacade.getProject()

I think so, and it's logged in some other issues that we may want to get rid of it to provide more clarity and flexibility (basically removing some apparent constraints).

> IMavenExecutionContext.getExecutionRequest() should contain a variant where one can pass a IMavenProjectFacade then which is then used to construct the appropiate Execution context without the need to store the basedir in the IMaven.

Again, that's what I started with, and it becomes a maze.

Reply:
We might then simply set the property here as a first start this should cover all relevant use-cases (see comment above).

Reply:
It's not "simply" and what you suggest is adding 1 more API when we can do without it.

Reply:
I don't know what new API you are referring here...  I'm refring to setting this in `ProjectRegistryManager.configureExecutionRequest(MavenExecutionRequest, IProjectRegistry, IFile, ResolverConfiguration)` this won't require new API and should suffice here.

Reply:
That's only 1 place where this API is used, and not sufficient. This method is only used in ProjectRegistryManager, some other places do require to have this (and more) set specifically for a given project, and that's for this kind of cases that we've created `MavenProjectFacade.getMaven()`

Reply:
> That's only 1 place where this API is used, and not sufficient.

I still don't see why it is not sufficient, can you please give a concrete code reference where it wont work?

> This method is only used in ProjectRegistryManager,

Yes and this are the places where current code starts the call, and all other code just "connects" there, because if there is no project context there is no MultiModule directory and that is the place where the project facade is managed! So I don't see where (and why) a project is actually executed if not through the ProjectRegistryManager?

Reply:
Also if you take a look at the stack trace of https://github.com/eclipse-m2e/m2e-core/issues/666 you will clearly see that this is processed through the maven builder that exactly setup the execution context in that way in the stack frame here:

at org.eclipse.m2e.core.internal.project.registry.ProjectRegistryManager.refresh(ProjectRegistryManager.java:506)

and that's where the issue arises not elsewhere in the code, because there are only a build of the project from that point and all other code simply attach to that execution afterwards!

Reply:
If we fix only the ProjectRegistryManager, then further errors of that sort will happen eg if AbstractLifecycleMapping is used because the remote repo wouldn't be resolved and some dependencies/plugins won't be properly resolved for the project at that stage (although they were resolved earlier).